projects/libs/core-library-typescript/src/virtual-machine/hard-coded-contracts/registry/kinatn-smart-contract.ts

```diff
@@ -0,0 +1,77 @@
+import BaseSmartContract from "../base-smart-contact";
+
+export default class KinAtnSmartContract extends BaseSmartContract {
+  static readonly ALLOCATED_POOL_STATE_VARIABLE = "atn_pool";
+  static readonly ONE_MILLION = 1_000_000;
+  static readonly ONE_BILLION = KinAtnSmartContract.ONE_MILLION * 1000;
+  static readonly DEFAULT_BALANCE = 10000;
+  static readonly POOL_INITIAL_VALUE = KinAtnSmartContract.ONE_BILLION * 20;
+
+  public async transfer(recipient: string, amount: number) {
+    if (amount <= 0) {
+      throw this.validationError("Transaction amount must be > 0");
+    }
+
+    let senderBalance: number = await this.getBalanceForAccount(this.senderAddressBase58);
+    // auto-fund accounts
+    if (senderBalance == 0) {
+      senderBalance = await this.financeAccount(this.senderAddressBase58);
+    }
+
+    if (senderBalance < amount) {
+      throw this.validationError(`Insufficient balance ${senderBalance} < ${amount}`);
+    }
+
+    let recipientBalance: number = await this.getBalanceForAccount(recipient);
+    if (recipientBalance == 0) {
+      // this is to finance a new account that is receiving money for the first time
+      recipientBalance = await this.financeAccount(recipient);
+    }
+
+    await this.setBalance(this.senderAddressBase58, senderBalance - amount);
+    await this.setBalance(recipient, recipientBalance + amount);
+  }
+
+  public async getBalance() {
+    return this.getBalanceForAccount(this.senderAddressBase58);
+  }
+
+  private async reduceFromPool(amount: number) {
+    let allocatedBalance = await this.getPoolBalance();
+    if (allocatedBalance == 0) {
+      // reset/init the pool
+      await this.setPoolBalance(KinAtnSmartContract.POOL_INITIAL_VALUE);
```

Comment:
If you reset the pool every time it's empty, what's the meaning of the balance? There's an infinite # of tokens